src/main/java/org/broadinstitute/hellbender/tools/walkers/fasta/ShiftFasta.java

```diff
@@ -0,0 +1,210 @@
+
+    public static final String SHIFT_OFFSET_LIST = "shift-offset-list";
+    @Argument(fullName = SHIFT_OFFSET_LIST,
+            doc="Number of bases to skip in the reference before starting the shifted reference. For example, if 300 is specified, the new fasta will start at the 301th base (count starting at 1)." +
+    "If not specified, the contig will be shifted by half the number of bases. To skip the shifting of a contig, specify 0 in the list.", optional = true)
+    private List<Integer> shiftOffsets = null;
+
+    public static final String INTERAL_FILE_NAME = "interval-file-name";
+    @Argument(fullName = INTERAL_FILE_NAME,
+            doc="Base name for interval files. Intervals will be midway between beginning and computed offset. If not specified or if custom offsets are specified, no interval files will be written.", optional = true)
+    private String intervalFilename;
+
+    public static final String LINE_WIDTH_LONG_NAME = "line-width";
+    @Argument(fullName= LINE_WIDTH_LONG_NAME, doc="Maximum length of sequence to write per line", optional=true)
+    public int basesPerLine = FastaReferenceWriter.DEFAULT_BASES_PER_LINE;
+
+    ReferenceDataSource refSource;
+    FastaReferenceWriter refWriter;
+    FileWriter chainFileWriter;
+    FileWriter intervalRegularWriter;
+    FileWriter intervalShiftedWriter;
+
+    int chainId = 0;
+
+    @Override
+    public boolean requiresReference() {
+        return true;
+    }
+
+    @Override
+    public void onTraversalStart() {
+        refSource = referenceArguments.getReferencePath() != null ? ReferenceDataSource.of(referenceArguments.getReferencePath()) : null;
+        final Path path = IOUtils.getPath(output);
+        chainId = 1;
+        try {
+            refWriter = new FastaReferenceWriterBuilder()
+                    .setFastaFile(path)
+                    .setBasesPerLine(basesPerLine)
+                    .build();
+            chainFileWriter = new FileWriter(shiftBackOutput);
+            if (intervalFilename != null) {
+                intervalRegularWriter = new FileWriter(intervalFilename+ ".intervals");
+                intervalShiftedWriter = new FileWriter(intervalFilename + ".shifted.intervals");
+            }
+        } catch (IOException e) {
+            throw new UserException.CouldNotCreateOutputFile("Couldn't create " + output + ", encountered exception: " + e.getMessage(), e);
+        }
+    }
+
+    public void traverse() {
+        SAMSequenceDictionary refDict = refSource.getSequenceDictionary();
+        long refLengthLong = refDict.getReferenceLength();
+        if (refLengthLong > Integer.MAX_VALUE) {
+            // TODO fix this??
+            throw new UserException.BadInput("Reference length is too long");
+        }
+        List<SAMSequenceRecord> contigs = refSource.getSequenceDictionary().getSequences();
+        final ListIterator<Integer> shiftOffsetsIt = (shiftOffsets != null && shiftOffsets.size() == contigs.size()) ?
```

Comment:
Can you put a check here that makes sure the list is the same size as the length of the fasta file?

Reply:
isn't that what this check is doing? making sure that we have a shift offset for each contig:
shiftOffsets.size() == contigs.size()

Reply:
You are right. However I still think we should be very careful about the shift offsets input. I think we should put a check here that asserts it is either empty OR it is exactly the lenght of the contigs in the file, otherwise if it is the wrong size we should throw a user exception since clearly there was a mistake.